internal/worker/changestreampruner/worker.go

```diff
@@ -66,91 +63,135 @@ func (c *WorkerConfig) Validate() error {
+	if err := cfg.Validate(); err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	pruner := &Pruner{
+	runner, err := worker.NewRunner(worker.RunnerParams{
+		Name:          "changestream-pruner",
+		IsFatal:       func(err error) bool { return false },
+		ShouldRestart: func(err error) bool { return false },
+		Clock:         cfg.Clock,
+		Logger:        internalworker.WrapLogger(cfg.Logger),
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &Pruner{
 		cfg:     cfg,
-		windows: make(map[string]window),
+		runner:  runner,
+		windows: make(map[string]Window),
 	}
 
-	pruner.tomb.Go(pruner.loop)
+	if err := catacomb.Invoke(catacomb.Plan{
+		Name: "changestream-pruner",
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{runner},
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
 
-	return pruner, nil
+	return w, nil
 }
 
 // Kill is part of the worker.Worker interface.
 func (w *Pruner) Kill() {
-	w.tomb.Kill(nil)
+	w.catacomb.Kill(nil)
 }
 
 // Wait is part of the worker.Worker interface.
 func (w *Pruner) Wait() error {
-	return w.tomb.Wait()
+	return w.catacomb.Wait()
+}
+
+// Report returns a map of internal state for debugging purposes.
+func (w *Pruner) Report() map[string]any {
+	return w.runner.Report()
 }
 
 func (w *Pruner) loop() error {
-	timer := w.cfg.Clock.NewTimer(defaultPruneMinInterval)
+	ctx := w.catacomb.Context(context.Background())
+
+	timer := w.cfg.Clock.NewTimer(defaultPruneInterval)
 	defer timer.Stop()
 
-	var attempts int
 	for {
 		select {
-		case <-w.tomb.Dying():
-			return tomb.ErrDying
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
 
 		case <-timer.Chan():
 			// Attempt to prune, if there is any critical error, kill the
 			// worker, which should force a restart.
-			pruned, err := w.prune()
-			if err != nil {
+			if err := w.prune(ctx); err != nil {
 				return errors.Trace(err)
 			}
-
-			// If nothing was pruned, increment the attempts counter, otherwise
-			// reset it. This should wind out the backoff strategy if there is
-			// nothing to prune, thus reducing the frequency of the pruner.
-			if len(pruned) == 0 {
-				attempts++
-			} else {
-				attempts = 0
-			}
-
-			timer.Reset(backOffStrategy(0, attempts))
+			timer.Reset(defaultPruneInterval)
 		}
 	}
 }
 
-func (w *Pruner) prune() (map[string]int64, error) {
-	ctx, cancel := w.scopedContext()
-	defer cancel()
+func (w *Pruner) prune(ctx context.Context) error {
+	modelNamespaces, err := w.getModelNamespaces(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	traceEnabled := w.cfg.Logger.IsLevelEnabled(logger.TRACE)
 	if traceEnabled {
-		w.cfg.Logger.Tracef(ctx, "Starting pruning change log")
+		w.cfg.Logger.Tracef(ctx, "pruning model namespaces: %v", modelNamespaces)
 	}
 
+	for _, mn := range modelNamespaces {
+		err := w.runner.StartWorker(ctx, mn.Namespace, func(ctx context.Context) (worker.Worker, error) {
+			db, err := w.cfg.DBGetter.GetDB(ctx, mn.Namespace)
```

Comment:
Should we not instead inject a service, and hide away the queries as implementation details?

Reply:
This is done now.